### PR TITLE
dead-code-lint: de-dup multiple unused assoc functions

### LINKED ISSUE #109600: Emit one warning when a struct has multiple unused associated functions
The change proposed in #97643 should also be applied to associated functions.

### PR DESCRIPTION
**Pull request by Ezrashaw**
Fixes #109600

Prior art: #97853

### CODE REVIEW

File: compiler/rustc_passes/src/dead.rs [hunk 1]
```diff
@@ -836,6 +836,13 @@ fn check_mod_deathness(tcx: TyCtxt<'_>, module: LocalDefId) {
     let module_items = tcx.hir_module_items(module);
 
     for item in module_items.items() {
+        if let hir::ItemKind::Impl(impl_item) = tcx.hir().item(item).kind {
+            for item in impl_item.items {
+                visitor.check_definition(item.id.owner_id.def_id);
+            }
+            continue;
```
Comment (lcnr) [inline compiler/rustc_passes/src/dead.rs:863]: this continue feels unnecessary, but also isn't really harmful :shrug:

File: compiler/rustc_passes/src/dead.rs [hunk 2]
```diff
@@ -837,9 +840,26 @@ fn check_mod_deathness(tcx: TyCtxt<'_>, module: LocalDefId) {
 
     for item in module_items.items() {
         if let hir::ItemKind::Impl(impl_item) = tcx.hir().item(item).kind {
+            let mut dead_items = Vec::new();
             for item in impl_item.items {
-                visitor.check_definition(item.id.owner_id.def_id);
+                match item.kind {
+                    hir::AssocItemKind::Const | hir::AssocItemKind::Type => {
+                        visitor.check_definition(item.id.owner_id.def_id)
```
Comment (lcnr) [inline compiler/rustc_passes/src/dead.rs:847]: would it make sense to do this for all associated items, and then change the message to "associated items are not used" if it isn't just methods?

**Review by lcnr — COMMENTED** [review]
nits, then r=me

**Comment by Ezrashaw** [comment]
@lcnr Implemented.

Note that this lint only applies to inherent `impl`s and therefore, associated types here are kind of useless without the incomplete `inherent_associated_types` feature.

**Comment by Ezrashaw** [comment]
@lcnr Sorry, I forgot to bless a single test... dammit.

On the other hand, that test's output has objectively improved.

**Pull request merged**